### PR TITLE
python support

### PR DESCRIPTION
**Pull request by yaronschwimmer**
- [X] Ready for review
- [X] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
Adds application scanning for Python (Poetry and Pip).

#### What are the relevant tickets?
https://snyksec.atlassian.net/browse/MYC-148
https://snyksec.atlassian.net/browse/MYC-149
https://snyksec.atlassian.net/browse/MYC-150

#### Screenshots
image
image

### CODE REVIEW

File: lib/analyzer/applications/python/poetry.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  for (const pathPair of filePairs) {
+    const depGraph = await lockFileParser.buildDepGraph(
+      filePathToContent[pathPair.manifest],
+      filePathToContent[pathPair.lock],
+      shouldIncludeDevDependencies,
+    );
+
+    const depGraphFact: DepGraphFact = {
+      type: "depGraph",
+      data: depGraph as DepGraph,
```
Comment (RotemS) [inline lib/analyzer/applications/python/poetry.ts:30]: Same as above - can this be undefined and is it ok :)

File: lib/analyzer/applications/python/pip.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+      metadataItems,
+    );
+    const depGraph = await builder.build();
+    if (!depGraph) {
+      continue;
+    }
+
+    const depGraphFact: DepGraphFact = {
+      type: "depGraph",
+      data: depGraph,
```
[the remark this answers is not in the document]
Comment (yaronschwimmer) [reply lib/analyzer/applications/python/pip.ts:116]: just a few lines above, there's
```javascript
if (!depGraph) {
  continue;
}
```
so we won't get here if `depGraph` is undefined

File: lib/analyzer/applications/python/poetry.ts (hunk 1 above)
Comment (yaronschwimmer) [reply lib/analyzer/applications/python/poetry.ts:30]: good catch. here I should add a check before like in the `pip` analyzer

File: lib/python-parser/metadata-parser.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  let name = "";
+  let version = "";
+  const dependencies: PythonRequirement[] = [];
+  for (let line of lines) {
+    line = line.trim();
+    if (line.length === 0) {
+      continue;
+    }
+    if (line.startsWith(PACKAGE_NAME)) {
+      name = line.substring(PACKAGE_NAME.length);
```
Comment (RotemS) [inline lib/python-parser/metadata-parser.ts:19]: Do you need it with the `Name` etc prefix or am I missing something? :D

File: lib/python-parser/metadata-parser.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      );
+      if (pythonPackage) {
+        dependencies.push(pythonPackage);
+      }
+    }
+  }
+  return { name: name.toLowerCase(), version, dependencies } as PythonPackage;
+}
+
+function parseDependency(dep: string): PythonRequirement | null {
```
Comment (RotemS) [inline lib/python-parser/metadata-parser.ts:34]: Is this the dependency name & version? Can we give the arg a more explanatory name?

File: test/lib/analyzer/python-pip-analyzer.spec.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
+import { readdirSync, readFileSync } from "fs";
+import * as path from "path";
+import { pipFilesToScannedProjects } from "../../../lib/analyzer/applications";
+import { FilePathToContent } from "../../../lib/analyzer/applications/types";
+
+function prepareFiles(testPath: string): FilePathToContent {
+  const basePath = path.join(__dirname, `../../fixtures/python/${testPath}`);
+  const filePathToContent = {
```
Comment (RotemS) [inline test/lib/analyzer/python-pip-analyzer.spec.ts:8]: Non blocking:
There are nice helpers for loading fixtures in `test/util/index.ts` that handle the path, `toString` etc

File: test/unit/python-requirements-parser.spec.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    const fileContent = `
+      jinja2==2.7.2
+      oauth2
+      pyramid~=1.3a6
+    `;
+    const requirements = getRequirements(fileContent);
+    expect(requirements).toHaveLength(3);
+    expect(requirements).toMatchObject([
+      { name: "jinja2", specifier: "=", version: "2.7.2" },
+      { name: "oauth2" },
```
Comment (RotemS) [inline test/unit/python-requirements-parser.spec.ts:29]: What happens in this case where we don't have a version?

**Review by RotemS — COMMENTED** [review]
Amazing work 💯 🥇 
Left some comments / questions :)

File: lib/python-parser/metadata-parser.ts (hunk 3 above)
Comment (yaronschwimmer) [reply lib/python-parser/metadata-parser.ts:19]: Not sure I understand the question. If the line starts with `Name: `, then the name of the package is what comes after that

File: test/unit/python-requirements-parser.spec.ts (hunk 6 above)
Comment (yaronschwimmer) [reply test/unit/python-requirements-parser.spec.ts:29]: https://github.com/snyk/snyk-docker-plugin/pull/443/files#diff-c2e9b021b0d027e8d361854e722e693c22bcc7f95a72cc146739d0a05d117826R64
We will take the latest version we find for this package. Or at least that was the intention, but looking at it I found a bug (the sorting comes after). I'll fix and add a test case.

File: lib/python-parser/metadata-parser.ts (hunk 3 above)
Comment (RotemS) [reply lib/python-parser/metadata-parser.ts:19]: Ignore, misread it :D

**Review by RotemS — APPROVED** [review]
🚀

**Pull request merged**